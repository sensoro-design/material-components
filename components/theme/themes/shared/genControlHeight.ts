import type { HeightMapToken, SeedToken } from '../../interface';

export const genControlHeight = (token: SeedToken): HeightMapToken => {
  const { controlHeight } = token;

  return {
    controlHeightSM: controlHeight * 0.75,
    controlHeightXS: controlHeight * 0.5,
    controlHeightLG: controlHeight * 1.25,
  };
};
