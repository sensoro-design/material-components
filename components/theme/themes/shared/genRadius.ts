import type { MapToken } from '../../interface';

type Radius = Pick<MapToken, 'borderRadiusSmall' | 'borderRadiusLarge' | 'borderRadiusRound' | 'borderRadius'>;

export const genRadius = (radiusBase: number): Radius => {
  let radiusSmall = radiusBase;
  let radiusLarge = radiusBase + 2;
  let radiusRound = radiusBase + 18;

  return {
    borderRadius: radiusBase > 16 ? 16 : radiusBase,
    borderRadiusSmall: radiusSmall,
    borderRadiusLarge: radiusLarge,
    borderRadiusRound: radiusRound,
  };
}
