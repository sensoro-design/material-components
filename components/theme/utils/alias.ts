import { getAlphaColor } from './getAlphaColor';
import { seedToken } from '../themes/seed';

import type { AliasToken, MapToken, OverrideToken, SeedToken } from '../interface';

/** Raw merge of `@ant-design/cssinjs` token. Which need additional process */
type RawMergedToken = MapToken & OverrideToken & { override: Partial<AliasToken> };

/**
 * Seed (designer) > Derivative (designer) > Alias (developer).
 *
 * Merge seed & derivative & override token and generate alias token for developer.
 */
export function formatToken(derivativeToken: RawMergedToken): AliasToken {
  const { override, ...restToken } = derivativeToken;
  const overrideTokens = { ...override };

  Object.keys(seedToken).forEach((token) => {
    delete overrideTokens[token as keyof SeedToken];
  });

  const mergedToken = {
    ...restToken,
    ...overrideTokens,
  };

  // Generate alias token
  const aliasToken: AliasToken = {
    ...mergedToken,

    paddingXXS: mergedToken.sizeXXS,
    paddingXS: mergedToken.sizeXS,
    paddingSM: mergedToken.sizeSM,
    padding: mergedToken.size,
    paddingMD: mergedToken.sizeMD,
    paddingLG: mergedToken.sizeLG,
    paddingXL: mergedToken.sizeXL,

    marginXXS: mergedToken.sizeXXS,
    marginXS: mergedToken.sizeXS,
    marginSM: mergedToken.sizeSM,
    margin: mergedToken.size,
    marginMD: mergedToken.sizeMD,
    marginLG: mergedToken.sizeLG,
    marginXL: mergedToken.sizeXL,
    marginXXL: mergedToken.sizeXXL,
  }

  return aliasToken;
}
