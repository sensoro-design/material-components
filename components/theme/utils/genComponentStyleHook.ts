import { useContext } from 'react';
import { useStyleRegister } from '@ant-design/cssinjs';
import { ConfigContext } from '../../config-provider/ConfigContext';
import { genCommonStyle, } from '../../style';
import { useToken, mergeToken, statisticToken } from '../internal';

import type { CSSInterpolation } from '@ant-design/cssinjs';
import type { ComponentTokenMap, GlobalToken, } from '../interface';
import type { UseComponentStyleResult } from '../internal';

export type OverrideTokenWithoutDerivative = ComponentTokenMap;
export type OverrideComponent = keyof OverrideTokenWithoutDerivative;

export type GlobalTokenWithComponent<ComponentName extends OverrideComponent> = GlobalToken &
  ComponentTokenMap[ComponentName];

export type TokenWithCommonCls<T> = T & {
  /** Wrap component class with `.` prefix */
  componentCls: string;
  /** Origin prefix which do not have `.` prefix */
  prefixCls: string;
  /** Wrap icon class with `.` prefix */
  iconCls: string;
  /** Wrap ant prefixCls class with `.` prefix */
  senCls: string;
};

export interface StyleInfo<ComponentName extends OverrideComponent> {
  hashId: string;
  prefixCls: string;
  rootPrefixCls: string;
  iconPrefixCls: string;
  overrideComponentToken: ComponentTokenMap[ComponentName];
}

export type FullToken<ComponentName extends OverrideComponent> = TokenWithCommonCls<
  GlobalTokenWithComponent<ComponentName>
>;

export function genComponentStyleHook<ComponentName extends OverrideComponent>(
  component: ComponentName,
  styleFn: (token: FullToken<ComponentName>, info: StyleInfo<ComponentName>) => CSSInterpolation,
  getDefaultToken?:
    | OverrideTokenWithoutDerivative[ComponentName]
    | ((token: GlobalToken) => OverrideTokenWithoutDerivative[ComponentName]),
) {
  return (prefixCls: string) => {
    const [theme, token, hashId] = useToken();
    const { getPrefixCls, iconPrefixCls } = useContext(ConfigContext);
    const rootPrefixCls = getPrefixCls();

    return [
      useStyleRegister(
        { theme, token, hashId, path: [component, prefixCls, iconPrefixCls] },
        () => {
          const { token: proxyToken, flush } = statisticToken(token);

          const defaultComponentToken = typeof getDefaultToken === 'function'
            ? getDefaultToken(proxyToken)
            : getDefaultToken;
          const mergedComponentToken = { ...defaultComponentToken, ...token[component] };

          const componentCls = `.${prefixCls}`;
          const mergedToken = mergeToken<
            TokenWithCommonCls<GlobalTokenWithComponent<OverrideComponent>>
          >(
            proxyToken,
            {
              componentCls,
              prefixCls,
              iconCls: `.${iconPrefixCls}`,
              senCls: `.${rootPrefixCls}`,
            },
            mergedComponentToken,
          );

          const styleInterpolation = styleFn(mergedToken as unknown as FullToken<ComponentName>, {
            hashId,
            prefixCls,
            rootPrefixCls,
            iconPrefixCls,
            overrideComponentToken: token[component],
          });
          flush(component, mergedComponentToken);

          return [styleInterpolation];
        }
      ),
      hashId,
    ]
  }
}
