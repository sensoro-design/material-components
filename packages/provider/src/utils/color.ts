import { TinyColor } from '@ctrl/tinycolor';

/**
 * 把一个颜色设置一下透明度
 * @example (#fff, 0.5) => rgba(255, 255, 255, 0.5)
 * @param baseColor {string}
 * @param alpha {0-1}
 * @returns rgba {string}
 */
export const setAlpha = (baseColor: string, alpha: number) =>
  new TinyColor(baseColor).setAlpha(alpha).toRgbString();

/**
 * 把一个颜色修改一些明度
 * @example (#000, 50) => #808080
 * @param baseColor {string}
 * @param brightness {0-100}
 * @returns hexColor {string}
 */
export const lighten = (baseColor: string, brightness: number) =>
  new TinyColor(baseColor).lighten(brightness).toHexString();
