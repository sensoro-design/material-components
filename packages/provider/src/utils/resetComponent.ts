import type { CSSObject } from '@ant-design/cssinjs';
import type { MaterialAliasToken } from '../types';

export const resetComponent = (token: MaterialAliasToken): CSSObject => ({
  boxSizing: 'border-box',
  margin: 0,
  padding: 0,
  color: token.colorText,
  fontSize: token.fontSize,
  lineHeight: token.lineHeight,
  listStyle: 'none',
});
