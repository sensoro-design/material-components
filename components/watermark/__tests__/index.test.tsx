import React from 'react';
import { Watermark } from '../index';
import { mountTest } from '../../../tests/shared/mountTest';
import { render } from '../../../tests/utils';

describe('Watermark', () => {
  const mockSrcSet = jest.spyOn(Image.prototype, 'src', 'set');

  beforeAll(() => {
    mockSrcSet.mockImplementation(function fn() {
      this.onload?.();
    });
  });

  afterAll(() => {
    mockSrcSet.mockRestore();
  });

  mountTest(Watermark);

  it('The watermark should render successfully', () => {
    const { container } = render(<Watermark className="watermark" content="Sensoro Design" />);
    expect(container.querySelector('.watermark div')).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('The offset should be correct', () => {
    const { container } = render(
      <Watermark
        className="watermark"
        offset={[200, 200]}
        content={['Sensoro Design', 'Sensoro Design Pro']}
      />,
    );

    const target = container.querySelector<HTMLDivElement>('.watermark div');

    expect(target?.style.left).toBe('150px');
    expect(target?.style.top).toBe('150px');
    expect(target?.style.width).toBe('calc(100% - 150px)');
    expect(target?.style.height).toBe('calc(100% - 150px)');
    expect(container).toMatchSnapshot();
  });
})
