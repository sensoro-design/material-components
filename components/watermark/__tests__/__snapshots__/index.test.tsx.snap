// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Watermark The offset should be correct 1`] = `
<div>
  <div
    class="watermark"
    style="position: relative;"
  >
    <div
      style="z-index: 9; position: absolute; left: 150px; top: 150px; width: calc(100% - 150px); height: calc(100% - 150px); pointer-events: none; background-repeat: repeat; background-position: 0px 0px; background-image: url('data:image/png;base64,00'); background-size: 236px;"
    />
  </div>
</div>
`;

exports[`Watermark The watermark should render successfully 1`] = `
<div>
  <div
    class="watermark"
    style="position: relative;"
  >
    <div
      style="z-index: 9; position: absolute; left: 0; top: 0; width: 100%; height: 100%; pointer-events: none; background-repeat: repeat; background-position: 0px 0px; background-image: url('data:image/png;base64,00'); background-size: 228px;"
    />
  </div>
</div>
`;
